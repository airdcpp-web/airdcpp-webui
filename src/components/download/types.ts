import * as UI from 'types/ui';


export type BrowseHandler = (() => void) | undefined;

export type PathDownloadHandler = (targetPath: string, targetFilename?: string) => Promise<any>;

export type LayoutProps = React.PropsWithChildren<{
  menuItems: React.ReactNode[];
  title: string;
  handleBrowse: BrowseHandler;
  t: UI.TranslateF;
}>;
