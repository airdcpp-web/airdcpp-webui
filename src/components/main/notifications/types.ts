export type NotificationLevel = 'error' | 'warning' | 'info' | 'success';
