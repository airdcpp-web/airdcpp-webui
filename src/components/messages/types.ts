import { IntersectionObserverProps } from 'react-intersection-observer';

import * as UI from '@/types/ui';

export interface CommonMessageProps extends Omit<IntersectionObserverProps, 'children'> {
  highlightMenuProps: UI.MessageActionMenuData;
}
