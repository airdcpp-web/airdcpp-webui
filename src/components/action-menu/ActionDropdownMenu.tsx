import * as React from 'react';

import SectionedDropdown, {
  SectionedDropdownProps,
} from '@/components/semantic/SectionedDropdown';
import MenuSection from '@/components/semantic/MenuSection';
import ActionMenuDecorator, {
  ActionMenuDecoratorChildProps,
} from '@/components/action-menu/decorators/ActionMenuDecorator';
import { buildMenu } from './builder/expandMenuBuilder';

export interface ActionMenuProps extends SectionedDropdownProps {
  header?: React.ReactNode;
}

const ActionDropdownMenu = ({
  header,
  children,
  ...other
}: ActionMenuProps & ActionMenuDecoratorChildProps) => (
  <SectionedDropdown {...other}>
    <MenuSection caption={header}>{children()}</MenuSection>
  </SectionedDropdown>
);

const ActionMenuDecorated = ActionMenuDecorator<ActionMenuProps, any, any>(
  ActionDropdownMenu,
  buildMenu,
);

export { ActionMenuDecorated as ActionMenu };
