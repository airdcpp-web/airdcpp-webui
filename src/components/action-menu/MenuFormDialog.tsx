import { useRef } from 'react';
import * as React from 'react';

import Modal, { ModalProps } from '@/components/semantic/Modal';
import Form, { FormSaveHandler } from '@/components/form/Form';

import * as UI from '@/types/ui';

import { IconType } from '@/components/semantic/Icon';
import { useTranslation } from 'react-i18next';
import { translate } from '@/utils/TranslationUtils';

export interface MenuFormDialogProps extends ModalProps {
  title: string;
  icon: IconType;
  fieldDefinitions: UI.FormFieldDefinition[];
  onSave: FormSaveHandler<UI.FormValueMap>;
}

export const MenuFormDialog: React.FC<MenuFormDialogProps> = ({
  fieldDefinitions,
  onSave,
  title,
  icon,
  ...other
}) => {
  const formRef = useRef<Form | null>(null);
  const { t } = useTranslation();
  return (
    <Modal
      {...other}
      title={title}
      onApprove={() => formRef.current!.save()}
      approveCaption={translate('Continue', t, UI.Modules.COMMON)}
      closable={false}
      icon={icon}
      dynamicHeight={true}
    >
      <Form ref={formRef} onSave={onSave} fieldDefinitions={fieldDefinitions} />
    </Modal>
  );
};
