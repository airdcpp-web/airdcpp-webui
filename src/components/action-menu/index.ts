export * from './ActionDropdownMenu';
export * from './DropdownMenu';
export * from '../icon/MenuIcon';
