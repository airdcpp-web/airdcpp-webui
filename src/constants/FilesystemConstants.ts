const MODULE_URL = 'filesystem';
export default {
  MODULE_URL: MODULE_URL,
  LIST_URL: MODULE_URL + '/list_items',
  DIRECTORY_URL: MODULE_URL + '/directory',
  DISK_INFO_URL: MODULE_URL + '/disk_info',

  LOCATION_DOWNLOAD: 'download',
  LOCATION_GENERIC: 'settings',
};
