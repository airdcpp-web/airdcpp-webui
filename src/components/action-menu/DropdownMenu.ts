import TableDropdown, { TableDropdownProps } from 'components/semantic/TableDropdown';

import ActionMenuDecorator, {
  ActionMenuDecoratorProps,
} from './decorators/ActionMenuDecorator';
import DownloadMenuDecorator, {
  DownloadMenuDecoratorProps,
} from './decorators/DownloadMenuDecorator';
import UserMenuDecorator, {
  UserMenuDecoratorProps,
} from './decorators/UserMenuDecorator';

import { DropdownProps } from 'components/semantic/Dropdown';

import * as UI from 'types/ui';
import { ActionMenuProps, ActionMenu } from './ActionDropdownMenu';
import { buildMenu } from './builder/slidingMenuBuilder';

type TableActionMenuDropdownProps = Omit<TableDropdownProps, 'children' | 'caption'>;

export type TableActionMenuProps<ItemDataT extends UI.ActionMenuItemDataValueType> =
  TableActionMenuDropdownProps & ActionMenuDecoratorProps<ItemDataT>;
export const TableActionMenu = ActionMenuDecorator<TableActionMenuDropdownProps, any>(
  TableDropdown,
  buildMenu,
);

export type UserMenuProps = UserMenuDecoratorProps & ActionMenuProps;
export const UserMenu = UserMenuDecorator<DropdownProps>(ActionMenu);

export type TableUserMenuProps = UserMenuDecoratorProps & TableActionMenuDropdownProps;
export const TableUserMenu =
  UserMenuDecorator<TableActionMenuDropdownProps>(TableActionMenu);

export type DownloadMenuProps<ItemDataT extends UI.DownloadableItemInfo> =
  DownloadMenuDecoratorProps<ItemDataT> & ActionMenuProps;

export const DownloadMenu = DownloadMenuDecorator<DropdownProps, any>(ActionMenu);

export type TableDownloadMenuProps<ItemDataT extends UI.DownloadableItemInfo> =
  DownloadMenuDecoratorProps<ItemDataT> & TableActionMenuDropdownProps;

export const TableDownloadMenu = DownloadMenuDecorator<TableActionMenuDropdownProps, any>(
  TableActionMenu,
);
