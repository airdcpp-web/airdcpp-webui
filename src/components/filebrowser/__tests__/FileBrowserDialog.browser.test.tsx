import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { renderDataNode } from '@/tests/render/test-renderers';

import * as API from '@/types/api';

import { fireEvent, waitFor } from '@testing-library/dom';
import FilesystemConstants from '@/constants/FilesystemConstants';
import { FilesystemListContentResponse } from '@/tests/mocks/api/filesystem';
import { MenuFormDialog } from '@/components/action-menu/MenuFormDialog';
import {
  createTestModalController,
  useModalButton,
} from '@/tests/helpers/test-dialog-helpers';
import {
  clickButton,
  waitExpectRequestToMatchSnapshot,
  waitForData,
} from '@/tests/helpers/test-helpers';
import {
  setInputFieldValues,
  setInputFieldValuesByPlaceholder,
  setupUserEvent,
} from '@/tests/helpers/test-form-helpers';
import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';
import { getMockServer, MockServer } from '@/tests/mocks/mock-server';

// tslint:disable:no-empty
describe('FileBrowserDialog', () => {
  let server: MockServer;

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  const getSocket = async (platform: API.PlatformEnum, disableLogging = false) => {
    const commonData = await initCommonDataMocks(server, {
      platform,
      socketOptions: { logLevel: disableLogging ? 'none' : 'warn' },
    });

    // Browse dialog
    server.addRequestHandler(
      'POST',
      FilesystemConstants.LIST_URL,
      FilesystemListContentResponse,
    );

    const onDirectoryCreated = vi.fn();
    server.addRequestHandler(
      'POST',
      FilesystemConstants.DIRECTORY_URL,
      undefined,
      onDirectoryCreated,
    );

    return { commonData, server, onDirectoryCreated };
  };

  interface RenderProps {
    fieldType: API.SettingTypeEnum;
    initialPath?: string;
    platform: API.PlatformEnum;
    disableLogging?: boolean;
  }

  const renderDialog = async ({
    fieldType,
    initialPath = '',
    platform,
    disableLogging,
  }: RenderProps) => {
    const { commonData, server, ...other } = await getSocket(platform, disableLogging);

    const onSave = vi.fn(() => Promise.resolve());
    const caption = 'Test dialog';

    const FileBrowserDialogTest = () => {
      const FormDefinitions: API.SettingDefinition[] = [
        {
          key: fieldType,
          type: fieldType,
          default_value: initialPath,
          title: fieldType,
        },
      ];

      const { open, onClose, button } = useModalButton();
      return (
        <>
          {button}
          {open && (
            <MenuFormDialog
              title={caption}
              onSave={onSave}
              icon="test"
              fieldDefinitions={FormDefinitions}
              onClose={onClose}
            />
          )}
        </>
      );
    };

    const renderData = renderDataNode(<FileBrowserDialogTest />, commonData);

    const userEvent = setupUserEvent();
    const modalController = createTestModalController({ ...renderData, userEvent });
    return { modalController, onSave, caption, userEvent, ...renderData, ...other };
  };

  describe('directory selection', () => {
    test('should select initial directory on linux', async () => {
      const { getByText, modalController, queryByText, getByRole, onSave, caption } =
        await renderDialog({
          fieldType: API.SettingTypeEnum.DIRECTORY_PATH,
          platform: API.PlatformEnum.LINUX,
        });

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      clickButton('Browse', getByRole);
      await waitForData('Loading items', queryByText);

      clickButton('Select', getByRole);

      // Save
      await modalController.closeDialogButton('Continue');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);

    test('should not select root on windows', async () => {
      const { getByText, modalController, queryByText, getByRole, /*onSave,*/ caption } =
        await renderDialog({
          fieldType: API.SettingTypeEnum.DIRECTORY_PATH,
          platform: API.PlatformEnum.WINDOWS,
        });

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      clickButton('Browse', getByRole);
      await waitForData('Loading items', queryByText);

      expect(getByText('Select').closest('button')).toHaveClass('disabled');
    }, 100000);

    test('should select child', async () => {
      const { getByText, modalController, queryByText, getByRole, onSave, caption } =
        await renderDialog({
          fieldType: API.SettingTypeEnum.DIRECTORY_PATH,
          platform: API.PlatformEnum.LINUX,
        });

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      clickButton('Browse', getByRole);
      await waitForData('Loading items', queryByText);

      expect(fireEvent.click(getByText('2 folders'))).toBeTruthy();

      clickButton('Select', getByRole);

      // Save
      await modalController.closeDialogButton('Continue');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);

    test('should create new directory', async () => {
      const renderResult = await renderDialog({
        fieldType: API.SettingTypeEnum.DIRECTORY_PATH,
        platform: API.PlatformEnum.LINUX,
      });
      const {
        getByText,
        modalController,
        queryByText,
        getByRole,
        onSave,
        caption,
        container,
        onDirectoryCreated,
        userEvent,
      } = renderResult;

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      clickButton('Browse', getByRole);
      await waitForData('Loading items', queryByText);

      // Create new directory
      expect(fireEvent.click(getByText('Create directory'))).toBeTruthy();

      await waitFor(() =>
        expect(container.querySelector('.ui.action.input.visible')).toBeTruthy(),
      );

      userEvent.click(getByText('Create directory'));
      // await setInputFieldValues(renderResult, { 'Create directory': 'New directory' });

      await setInputFieldValuesByPlaceholder(renderResult, {
        'Directory name': 'New directory',
      });

      clickButton('Create', getByRole);
      await waitFor(() => {
        expect(onDirectoryCreated).toHaveBeenCalledTimes(1);
      });

      await waitFor(() => {
        expect(container.querySelector('.ui.action.input.visible')).toBeFalsy();
      });

      // Select the new directory
      clickButton('Select', getByRole);

      // Save
      await modalController.closeDialogButton('Continue');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);
  });

  describe('file selection', () => {
    test('should select existing', async () => {
      const { getByText, modalController, queryByText, getByRole, onSave, caption } =
        await renderDialog({
          fieldType: API.SettingTypeEnum.EXISTING_FILE_PATH,
          platform: API.PlatformEnum.LINUX,
        });

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      clickButton('Browse', getByRole);
      await waitForData('Loading items', queryByText);

      // Go to child directory
      expect(fireEvent.click(getByText('2 folders'))).toBeTruthy();
      await waitForData('Loading items', queryByText);

      // Select a file
      clickButton('empty.txt', getByRole);

      // Save
      await modalController.closeDialogButton('Continue');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);

    test('should select new', async () => {
      const userEvent = setupUserEvent();
      const {
        getByText,
        getByRole,
        modalController,
        queryByText,
        getByLabelText,
        onSave,
        caption,
      } = await renderDialog({
        fieldType: API.SettingTypeEnum.FILE_PATH,
        platform: API.PlatformEnum.LINUX,
      });

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      clickButton('Browse', getByRole);
      await waitForData('Loading items', queryByText);

      // Go to child directory
      expect(fireEvent.click(getByText('2 folders'))).toBeTruthy();
      await waitForData('Loading items', queryByText);

      // Enter name for the file
      await setInputFieldValues(
        { userEvent, getByLabelText },
        { Filename: 'new_file.txt' },
      );
      await userEvent.click(getByText('Select'));

      // Save
      await modalController.closeDialogButton('Continue');

      expect(onSave).toHaveBeenCalledTimes(1);
      expect(onSave.mock.calls[0]).toMatchSnapshot();
    }, 100000);
  });

  describe('errors', () => {
    const testErrorPlatform = async (errorPath: string, platform: API.PlatformEnum) => {
      const { getByText, modalController, getByRole, /*, onSave,*/ caption } =
        await renderDialog({
          fieldType: API.SettingTypeEnum.DIRECTORY_PATH,
          platform,
          initialPath: errorPath,
          disableLogging: true,
        });

      const onSuccessListContent = vi.fn();
      const onFailListContent = vi.fn(() => {
        // Set up success response for retries
        server.addRequestHandler(
          'POST',
          FilesystemConstants.LIST_URL,
          FilesystemListContentResponse,
          onSuccessListContent,
        );
      });

      // Set up error response for the initial loading
      server.addErrorHandler(
        'POST',
        FilesystemConstants.LIST_URL,
        "Directory '/non/existing/path' does not exist",
        400,
        onFailListContent,
      );

      await modalController.openDialog();

      // Check content
      await waitFor(() => expect(getByText(caption)).toBeTruthy());

      // Open dialog
      clickButton('Browse', getByRole);

      await waitExpectRequestToMatchSnapshot(onFailListContent);
      await waitExpectRequestToMatchSnapshot(onSuccessListContent);
    };

    test('should handle non-existing directories on initial loading on linux', async () => {
      const errorPath = '/non/existing/path/';
      await testErrorPlatform(errorPath, API.PlatformEnum.LINUX);
    }, 100000);

    test('should handle non-existing directories on initial loading on windows', async () => {
      const errorPath = 'C:\\non\\existing\\path\\';
      await testErrorPlatform(errorPath, API.PlatformEnum.WINDOWS);
    }, 100000);
  });
});
