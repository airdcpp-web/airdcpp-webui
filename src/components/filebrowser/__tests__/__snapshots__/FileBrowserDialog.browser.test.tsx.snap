// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileBrowserDialog > directory selection > should create new directory 1`] = `
[
  {
    "directory_path": "/New directory/",
  },
  {
    "directory_path": "/New directory/",
  },
]
`;

exports[`FileBrowserDialog > directory selection > should select child 1`] = `
[
  {
    "directory_path": "/2 folders/",
  },
  {
    "directory_path": "/2 folders/",
  },
]
`;

exports[`FileBrowserDialog > directory selection > should select initial directory on linux 1`] = `
[
  {
    "directory_path": "/",
  },
  {
    "directory_path": "/",
  },
]
`;

exports[`FileBrowserDialog > errors > should handle non-existing directories on initial loading on linux 1`] = `
{
  "data": {
    "directories_only": false,
    "path": "/non/existing/path/",
  },
  "method": "POST",
  "path": "filesystem/list_items",
}
`;

exports[`FileBrowserDialog > errors > should handle non-existing directories on initial loading on linux 2`] = `
{
  "data": {
    "directories_only": false,
    "path": "/",
  },
  "method": "POST",
  "path": "filesystem/list_items",
}
`;

exports[`FileBrowserDialog > errors > should handle non-existing directories on initial loading on windows 1`] = `
{
  "data": {
    "directories_only": false,
    "path": "C:\\non\\existing\\path\\",
  },
  "method": "POST",
  "path": "filesystem/list_items",
}
`;

exports[`FileBrowserDialog > errors > should handle non-existing directories on initial loading on windows 2`] = `
{
  "data": {
    "directories_only": false,
    "path": "",
  },
  "method": "POST",
  "path": "filesystem/list_items",
}
`;

exports[`FileBrowserDialog > file selection > should select existing 1`] = `
[
  {
    "existing_file_path": "/2 folders/empty.txt",
  },
  {
    "existing_file_path": "/2 folders/empty.txt",
  },
]
`;

exports[`FileBrowserDialog > file selection > should select new 1`] = `
[
  {
    "file_path": "/2 folders/new_file.txt",
  },
  {
    "file_path": "/2 folders/new_file.txt",
  },
]
`;
